proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,241 @@
+expects:
+
+- Data: A `u8` bit mask of the staged features.
+- Accounts:
+  - Staged Features PDA: writable
+  - Vote account: signer
+
+A bit mask is used as a compressed ordered list of indices. This has two main
+benefits:
+
+- Minimizes transaction size for nodes, requiring only one byte to describe
+  256 bytes (8 * 32) of data. The alternative would be sending `n` number of
+  32-byte addresses in each instruction.
+- Reduce compute required to search the list of staged features for a matching
+  address. The bitmask's format provides the Feature Gate program with the
+  indices it needs to store supported stake without searching through the staged
+  features for a match.
+
+A `1` bit represents support for a feature. For example, for staged features
+`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
+except `E` and `H`, their `u8` value would be 246, or `11110110`.
+
+A node's submitted bit mask is then used to add that node's current epoch stake
+to the Staged Features PDA for each feature for which it has signaled support.
+This is done by using the Get Epoch Stake Syscall.
+
+Nodes are required to submit a transaction containing this instruction:
+
+- Any time at least 4500 slots before the end of the epoch.
+- During startup after any reboot.
+
+Transactions sent too late (< 4500 slots before the epoch end) will be rejected
+by the Feature Gate program.
+
+If a node does not send this transaction or it is rejected, their stake is not
+tallied. This is analogous to a node sending this transaction in a valid point
+in the epoch signalling support for zero features.
+
+If a feature is revoked, the list of staged features will not change, and nodes
+may still signal support for this feature. However, the runtime will not
+activate this feature if its corresponding feature account no longer exists
+on-chain.
+
+### Step 4: Feature Activation
+
+During the epoch rollover, the runtime can simply load the Staged Features PDA
+and calculate the stake - as a percentage of the total epoch stake - in support
+for each feature ID to determine which staged features to activate.
+
+Only features whose stake support meets the required threshold are activated.
+This threshold will be hard-coded in the runtime to 95% initially, but future
+iterations on the process could allow feature key-holders to set a custom
+threshold per-feature.
+
+As mentioned previously, if a feature was revoked, it will no longer exist
+on-chain, and therefore will be not activated by the runtime, regardless of
+calculated stake support.
+
+If a feature is not activated, either because it has been revoked or it did not
+meet the required stake support, it must be resubmitted according to Step 2.
+
+Once the epoch rollover is complete, as soon as the Feature Gate program is
+invoked again - either via `StageFeatureForActivation` or
+`SignalSupportForStagedFeature` - the Staged Features PDA's account state is
+updated and the process begins again.
```

Comment:
Please detail what it means for the Staged Features PDA's account state to be
updated.